fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/ImplementRecursiveDfsUnionRule.java

```diff
@@ -0,0 +1,124 @@
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RecordQueryPlanMatchers.tempTableScanPlan;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.ReferenceMatchers.exploratoryMember;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RelationalExpressionMatchers.recursiveUnionExpression;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RelationalExpressionMatchers.selectExpression;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RelationalExpressionMatchers.tempTableInsertExpression;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.SetMatcher.exactlyInAnyOrder;
+
+@SuppressWarnings("PMD.TooManyStaticImports")
+public class ImplementRecursiveDfsUnionRule extends ImplementationCascadesRule<RecursiveUnionExpression> {
+
+    @Nonnull
+    private static final BindingMatcher<RecordQueryPlan> initialInnerPlanMatcher = anyPlan();
+
+    @Nonnull
+    private static final BindingMatcher<TempTableInsertPlan> initialPlanMatcher = tempTableInsertPlan(initialInnerPlanMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<PlanPartition> innerPlanPartitionMatcher = anyPlanPartition();
+
+    @Nonnull
+    private static final BindingMatcher<Collection<PlanPartition>> innerPlanPartitionsMatcher = rollUpPartitions(any(innerPlanPartitionMatcher));
+
+    @Nonnull
+    private static final BindingMatcher<Reference> innerReferenceMatcher = planPartitions(innerPlanPartitionsMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> innerQunMatcher = forEachQuantifierOverRef(innerReferenceMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> mapTempTableQunMatcher = forEachQuantifier(mapPlan(tempTableScanPlan()));
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> tempTableQunMatcher = forEachQuantifier(tempTableScanPlan());
+
+    @Nonnull
+    private static final BindingMatcher<SelectExpression> recursiveInnerSelectMatcher = selectExpression(
+            exactlyInAnyOrder(innerQunMatcher, tempTableQunMatcher.or(mapTempTableQunMatcher)));
+
+    @Nonnull
+    private static final BindingMatcher<TempTableInsertExpression> recursivePlanMatcher = tempTableInsertExpression(
+            forEachQuantifierOverRef(exploratoryMember(recursiveInnerSelectMatcher)));
+
+    @Nonnull
+    private static final BindingMatcher<RecursiveUnionExpression> root = recursiveUnionExpression(forEachQuantifier(initialPlanMatcher),
+            forEachQuantifier(recursivePlanMatcher)).where(preOrderTraversalIsAllowed());
+
+    public ImplementRecursiveDfsUnionRule() {
+        super(root);
+    }
+
+    @Override
+    public void onMatch(@Nonnull final ImplementationCascadesRuleCall call) {
+        final var recursiveUnion = call.get(root);
```

Comment:
The naming should be indicative what entity it is they are holding; this should be `recursiveUnionExpression`. `...Alias` is fine. In a ton of other code we say `...Quantifier` and `...Reference` as opposed to use abbreviations.